### ISSUE TITLE
cql_server - exception while processing connection on several shards if no space left on device

### ISSUE THREAD

**Issue by aleksbykov** [scylladb/scylla#6348]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash):
4.0.rc3-0.20200501.eee4c00e29 with build-id 0fe095b80ee54c7f47a688aaac0c6bb835118ddc
Cluster size: 3
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0352ad86d128c6419

Enospc nemesis, simulate  no space on device for node where it is running. 
The nemesis was running on the node on cluster. Scylla detect no space left on defice.

Next error found in db node log:

```
2020-05-01T06:30:45+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | sudo: pam_unix(sudo:session): session closed for user root
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !WARNING | scylla: [shard 1] storage_service - Shutting down communications due to I/O errors until operator intervention
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !WARNING | scylla: [shard 1] storage_service - Commitlog error: std::filesystem::__cxx11::filesystem_error (error system:28, filesystem error: open failed: No space left on device [/var/lib/scylla/hints/1/10.0.248.66/HintsLog-1-18014398509763643.log])
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !WARNING | scylla: [shard 1] commitlog - Exception in segment reservation: storage_io_error (Storage I/O error: 28: filesystem error: open failed: No space left on device [/var/lib/scylla/hints/1/10.0.248.66/HintsLog-1-18014398509763643.log])
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: [shard 0] storage_service - Stop transport: starts
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: [shard 0] storage_service - Thrift server stopped
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: [shard 12] cql_server - exception while processing connection: std::nested_exception
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: [shard 1] cql_server - exception while processing connection: std::system_error (error system:32, sendmsg: Broken pipe)
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: [shard 1] cql_server - exception while processing connection: std::nested_exception
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: [shard 0] storage_service - CQL server stopped
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: [shard 0] storage_service - Stop transport: shutdown rpc and cql server done
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: scylla: ./seastar/src/core/file.cc:492: virtual seastar::append_challenged_posix_file_impl::~append_challenged_posix_file_impl(): Assertion `_closing_state == state::closed' failed.
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: scylla: ./seastar/src/core/file.cc:492: virtual seastar::append_challenged_posix_file_impl::~append_challenged_posix_file_impl(): Assertion `_closing_state == state::closed' failed.
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: Aborting on shard 10.
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: Backtrace:
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000002ed8bc2
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000002e7e73b
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000002e7ea35
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000002e7ead0
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x00007f6e57d71b1f
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: /opt/scylladb/libreloc/libc.so.6+0x000000000003c624
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: /opt/scylladb/libreloc/libc.so.6+0x00000000000258d8
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: /opt/scylladb/libreloc/libc.so.6+0x00000000000257a8
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: /opt/scylladb/libreloc/libc.so.6+0x0000000000034a65
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000002e0a201
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000002e0a24c
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000000e39738
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000002e22881
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000000e58a94
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x00000000011cb0c4
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x000000000128ce45
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000001190a95
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000001190bcf
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x0000000001191846
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: 0x000000000313b4fc
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: [shard 0] gossip - My status = NORMAL
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: [shard 0] gossip - Announcing shutdown
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: [shard 0] storage_service - Node 10.0.25.159 state jump to normal
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !ERR     | systemd-coredump: Not enough disk space for coredump of 3433 (scylla), refusing.
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !CRIT    | systemd-coredump: Process 3433 (scylla) of user 996 dumped core.
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !NOTICE  | systemd: scylla-server.service: main process exited, code=killed, status=6/ABRT
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd: Stopped Run Scylla Housekeeping daily mode.
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd: Stopped Run Scylla Housekeeping restart mode.
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !NOTICE  | systemd: Unit scylla-server.service entered failed state.
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !WARNING | systemd: scylla-server.service failed.
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd: scylla-server.service holdoff time over, scheduling restart.
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd: Stopping Scylla JMX...
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !NOTICE  | systemd: scylla-jmx.service: main process exited, code=exited, status=143/n/a
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd: Stopped Scylla JMX.
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !NOTICE  | systemd: Unit scylla-jmx.service entered failed state.
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !WARNING | systemd: scylla-jmx.service failed.
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd: Stopped Scylla Server.
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd: Starting Scylla Server...
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | sshd[67130]: Accepted publickey for centos from 10.0.246.104 port 49212 ssh2: RSA SHA256:g7OmDbWC8qLUZHxjMohQe3LsQHfcOJbG1ZHqodB6smc
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd-logind: New session 1716 of user centos.
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd: Started Session 1716 of user centos.
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | sshd[67130]: pam_unix(sshd:session): session opened for user centos by (uid=0)
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | sshd[67130]: pam_unix(sshd:session): session closed for user centos
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd-logind: Removed session 1716.
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd: Stopping irqbalance daemon...
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd: Stopped irqbalance daemon.
2020-05-01T06:30:49+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | systemd: Started irqbalance daemon.
2020-05-01T06:30:50+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla_prepare: File /etc/sysconfig/irqbalance.scylla.orig already exists - not overwriting.
2020-05-01T06:30:50+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla_prepare: Restarting irqbalance: going to ban the following IRQ numbers: 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 150, 154, 147, 153, 149, 148, 151, 152 ...
2020-05-01T06:30:50+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla_prepare: Restarting irqbalance via systemctl...
2020-05-01T06:30:50+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla_prepare: No non-NVMe disks to tune
2020-05-01T06:30:50+00:00  longevity-10gb-3h-4-0-db-nod
```
Decoded backtrace:
```
2020-05-01T06:30:48+00:00  longevity-10gb-3h-4-0-db-node-1f528fb5-2 !INFO    | scylla: [shard 1] cql_server - exception while processing connection: std::nested_exception

void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at /usr/include/fmt/format.h:1158
seastar::backtrace_buffer::append_backtrace() at /usr/include/fmt/format.h:1158
 (inlined by) print_with_backtrace at ./build/release/seastar/./seastar/src/core/reactor.cc:755
seastar::print_with_backtrace(char const*) at /usr/include/fmt/format.h:1158
sigabrt_action at /usr/include/fmt/format.h:1158
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3427
 (inlined by) _FUN at ./build/release/seastar/./seastar/src/core/reactor.cc:3423
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
seastar::append_challenged_posix_file_impl::~append_challenged_posix_file_impl() at ./build/release/seastar/./seastar/include/seastar/core/future.hh:372
 (inlined by) seastar::append_challenged_posix_file_impl::~append_challenged_posix_file_impl() at ./build/release/seastar/./seastar/src/core/file.cc:489
non-virtual thunk to seastar::append_challenged_posix_file_impl::~append_challenged_posix_file_impl() at ./build/release/seastar/./seastar/include/seastar/core/future.hh:372
seastar::shared_ptr<seastar::file_impl>::~shared_ptr() at /usr/include/fmt/format.h:1158
 (inlined by) seastar::file::~file() at ././seastar/include/seastar/core/file.hh:157
 (inlined by) checked_file_impl::~checked_file_impl() at ./checked-file-impl.hh:30
 (inlined by) seastar::shared_ptr_count_for<checked_file_impl>::~shared_ptr_count_for() at ././seastar/include/seastar/core/shared_ptr.hh:463
 (inlined by) seastar::shared_ptr_count_for<checked_file_impl>::~shared_ptr_count_for() at ././seastar/include/seastar/core/shared_ptr.hh:463
seastar::shared_ptr<seastar::file_impl>::~shared_ptr() at ./build/release/seastar/./seastar/include/seastar/core/future.hh:1675
 (inlined by) seastar::file::~file() at ./build/release/seastar/./seastar/include/seastar/core/file.hh:157
 (inlined by) seastar::file_data_sink_impl::~file_data_sink_impl() at ./build/release/seastar/./seastar/src/core/fstream.cc:312
 (inlined by) seastar::file_data_sink_impl::~file_data_sink_impl() at ./build/release/seastar/./seastar/src/core/fstream.cc:312
std::default_delete<seastar::data_sink_impl>::operator()(seastar::data_sink_impl*) const at /usr/include/fmt/format.h:1158
 (inlined by) std::unique_ptr<seastar::data_sink_impl, std::default_delete<seastar::data_sink_impl> >::~unique_ptr() at /usr/include/c++/9/bits/unique_ptr.h:292
 (inlined by) seastar::data_sink::~data_sink() at ././seastar/include/seastar/core/iostream.hh:93
 (inlined by) seastar::output_stream<char>::~output_stream() at ././seastar/include/seastar/core/iostream.hh:308
sstables::file_writer::~file_writer() at /usr/include/fmt/format.h:1158
 (inlined by) sstables::sstable::write_toc(seastar::io_priority_class const&) at ./sstables/sstables.cc:957
sstables::mc::writer::writer(sstables::sstable&, schema const&, unsigned long, sstables::sstable_writer_config const&, encoding_stats, seastar::io_priority_class const&, unsigned int) at ././sstables/types.hh:261
 (inlined by) std::_MakeUniq<sstables::mc::writer>::__single_object std::make_unique<sstables::mc::writer, sstables::sstable&, schema const&, unsigned long&, sstables::sstable_writer_config const&, encoding_stats&, seastar::io_priority_class const&, unsigned int&>(sstables::sstable&, schema const&, unsigned long&, sstables::sstable_writer_config const&, encoding_stats&, seastar::io_priority_class const&, unsigned int&) at /usr/include/c++/9/bits/unique_ptr.h:857
 (inlined by) sstables::mc::make_writer(sstables::sstable&, schema const&, unsigned long, sstables::sstable_writer_config const&, encoding_stats, seastar::io_priority_class const&, unsigned int) at ./sstables/mc/writer.cc:1434
sstables::sstable_writer::sstable_writer(sstables::sstable&, schema const&, unsigned long, sstables::sstable_writer_config const&, encoding_stats, seastar::io_priority_class const&, unsigned int) at /usr/include/fmt/format.h:1158
sstables::sstable::get_writer(schema const&, unsigned long, sstables::sstable_writer_config const&, encoding_stats, seastar::io_priority_class const&, unsigned int) at /usr/include/fmt/format.h:1158
 (inlined by) operator() at ./sstables/sstables.cc:2465
apply at /usr/include/fmt/format.h:1158
 (inlined by) apply<sstables::sstable::write_components(flat_mutation_reader, uint64_t, sstables::schema_ptr, const sstables::sstable_writer_config&, encoding_stats, const seastar::io_priority_class&)::<lambda()> > at ././seastar/include/seastar/core/apply.hh:44
 (inlined by) apply_tuple<sstables::sstable::write_components(flat_mutation_reader, uint64_t, sstables::schema_ptr, const sstables::sstable_writer_config&, encoding_stats, const seastar::io_priority_class&)::<lambda()> > at ././seastar/include/seastar/core/future.hh:1618
 (inlined by) apply<sstables::sstable::write_components(flat_mutation_reader, uint64_t, sstables::schema_ptr, const sstables::sstable_writer_config&, encoding_stats, const seastar::io_priority_class&)::<lambda()> > at ././seastar/include/seastar/core/future.hh:1652
 (inlined by) operator() at ././seastar/include/seastar/core/thread.hh:253
 (inlined by) call at ././seastar/include/seastar/util/noncopyable_function.hh:101
seastar::noncopyable_function<void ()>::operator()() const at ./build/release/seastar/./seastar/include/seastar/util/noncopyable_function.hh:184
 (inlined by) seastar::thread_context::main() at ./build/release/seastar/./seastar/src/core/thread.cc:286

```
Coredump was saved, because of no space left on device

**Comment by fruch** [comment]
exact same thing happen in one of alternator runs, the core dump took place during shutdown.

but after that it did managed to boot with this failure, Startup failed: std::out_of_range (attempted to read 4 bytes from a 0 byte buffer), seem like the commitlog too corrupted and failed reading them:

```
2020-05-06T01:40:31+00:00  alternator-3h-master-db-node-dda90de6-6 !WARNING | scylla: [shard 12] commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-216172782117418478.log. 6664184 bytes skipped.
2020-05-06T01:40:32+00:00  alternator-3h-master-db-node-dda90de6-6 !WARNING | scylla: [shard 9] commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-162129586588972523.log. 27234296 bytes skipped.
2020-05-06T01:40:32+00:00  alternator-3h-master-db-node-dda90de6-6 !WARNING | scylla: [shard 13] commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-234187180626900461.log. 25288696 bytes skipped.
2020-05-06T01:40:33+00:00  alternator-3h-master-db-node-dda90de6-6 !WARNING | scylla: [shard 10] commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-180143985098454510.log. 27004920 bytes skipped.
2020-05-06T01:40:33+00:00  alternator-3h-master-db-node-dda90de6-6 !WARNING | scylla: [shard 3] commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-54043195532080625.log. 29368312 bytes skipped.
2020-05-06T01:40:34+00:00  alternator-3h-master-db-node-dda90de6-6 !WARNING | scylla: [shard 0] commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-3634669.log. 11309048 bytes skipped.
...
2020-05-06T01:40:46+00:00  alternator-3h-master-db-node-dda90de6-6 !ERR     | scylla: [shard 0] init - Startup failed: std::out_of_range (attempted to read 4 bytes from a 0 byte buffer)
```
* https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-alternator-3h-test/38/
* https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-alternator-3h-test/38/artifact/scylla-cluster-tests/latest/alternator-3h-master-db-cluster-dda90de6/alternator-3h-master-db-node-dda90de6-6/database.log/*view*/

**Comment by fruch** [comment]
@slivne @bhalevy  we are seeing those coredump during shutdown while at fulldisk for a while now, but it's the first time I've noticed it doesn't boot after that with such failure.

**Comment by slivne** [comment]
@bhalevy rings a bell ?

**Comment by bhalevy** [comment]
> @bhalevy rings a bell ?

yes, it's essentially a duplicate of #5506 (fix wip in #5509)

**Comment by slivne** [comment]
closing
